File: challenge_19/challenge_19.exs
defmodule Sundays do
  def count(final,[year|[_|[count|_]]],_) when year == final do
    IO.inspect count+1
  end

  def count(final,[year|[day|count]],months) when year < final do
    count(final,calendar(year,months,day,List.first(count)),months)
  end

  def calendar(year,months,day,count) when length(months) <= 0 do
    [year+1, day, count]
  end

  def calendar(year,months,day,count) when length(months) > 0 do
    [month|rem_months] = months

    if day == 0, do: count = count + 1
    if month == 28 and leap_year?(year), do: month = 29

    calendar(year,rem_months,(rem((month + day),7)),count)
  end

  def leap_year?(year) do
    unless((rem(year,100) == 0 and rem(year,400) != 0), do: rem(year,4) == 0)
  end
end

Sundays.count(2000,[1901,2,0],[31,28,31,30,31,30,31,31,30,31,30,31])

File: challenge_19/challenge_19.c
#include <stdio.h>

int main(void) {
  int year;
  int sundays       = 0;
  int current_date  = 1;
  int current_month = 0;
  int month[12]     = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };

  for (year = 1900; year < 2001; year++) {
    for (current_month = 0; current_month <= 11; current_month++) {
      if (current_date % 7 == 0 && year != 1900) {
        sundays += 1;
        current_date = 0;
      }

      if (current_month != 1) {
        current_date += month[current_month];
      } else {
        if (year % 4 == 0) {
          if (year % 100 != 0 || year % 400 == 0) {
            current_date += 29;
          } else {
            current_date += month[current_month];
          }
        } else {
          current_date   += month[current_month];
        }
      }
    }
  }

  printf("%d\n", sundays);
}
